src/lib/redirects-manager-context.tsx

```diff
@@ -0,0 +1,116 @@
+    wildcard: boolean;
+    saving?: boolean;
+    lastSave?: Date;
+    lastModification?: Date;
+    filterBy?: string;
+    perPage?: number;
+    currentPage?: number;
+    imported?: boolean;
+}
+
+export interface RedirectionProps {
+    id: string;
+    request?: string;
+    destination?: string;
+    modificationDate?: any;
+    warningRequestDuplication?: boolean;
+}
+
+export interface RedirectionsStore extends Array<RedirectionProps> { }
+
+type Dispatch = (action: Action) => void;
+interface RedirectsManagerProviderProps { children: React.ReactNode; }
+
+const RedirectsManagerContext = React.createContext<RedirectsManagerContextInterface | undefined>(undefined);
+const RedirectsManagerDispatchContext = React.createContext<Dispatch | undefined>(undefined);
+
+interface UpdateServerStateProps {dispatch: Dispatch; state: RedirectsManagerContextInterface; }
+
+export const updateServerState = async ({dispatch, state}: UpdateServerStateProps) => {
+    dispatch({type: 'saving-state', value: true});
+    try {
+        const newState = await saveState(state);
+        dispatch({type: 'set', value: {...newState, lastSave: new Date(), saving: false}});
+    } catch (e) {
+        dispatch({type: 'saving-state', value: false});
+    }
+};
+
+type Action = {type: 'saving-state', value: boolean} |
+              {type: 'add', value: null} |
+              {type: 'remove', value: RedirectionProps} |
+              {type: 'edit', value: RedirectionProps} |
+              {type: 'set', value: RedirectsManagerContextInterface};
+
+const redirectsManagerReduducer = (state: RedirectsManagerContextInterface, action: Action) => {
+    switch (action.type) {
+
+        case 'add': {
+            state.store.push({id: v4(), modificationDate: 'not saved'});
+            return {...state, currentPage: 1};
+        }
+
+        case 'edit': {
+            const index = state.store.findIndex((el) => {
+                return el.id === action.value.id;
+            });
+            state.store[index] = action.value;
+            state.lastModification = new Date();
+            return {...checkRepeatedRequests(state)};
+        }
+
+        case 'remove': {
+            state.store = state.store.filter((el) => {
+                return el.id !== action.value.id;
+            });
+            state.lastModification = new Date();
+            return {...checkRepeatedRequests(state)};
+        }
+
+        case 'set': {
+            return action.value;
+        }
+
+        case 'saving-state': {
+            state.saving = action.value;
+            return {...state};
+        }
+
+        default: {
+            throw new Error(`Unhandled action type`);
+        }
+    }
+};
+
+export const RedirectsManagerProvider = ({children}: RedirectsManagerProviderProps) => {
+    const [state, dispatch] = React.useReducer(redirectsManagerReduducer, {store: [], wildcard: false} );
+    return (
+        <RedirectsManagerContext.Provider value={state}>
```

Comment:
as you've already got full state-management capability from `useReducer`, there's no reason to rely on implicit dependencies introduced by the context. The state can simply be passed down as props. This will make all interfaces more explicit and clear.

Reply:
applied. Context was removed